src/scanpy/preprocessing/_scale.py

```diff
@@ -29,20 +29,32 @@
     da = None
 
 if TYPE_CHECKING:
-    from numpy.typing import NDArray
+    from typing import TypeVar
+
+    from numpy.typing import ArrayLike, NDArray
 
     from .._utils import _CSMatrix
 
+    _A = TypeVar("_A", bound=_CSMatrix | np.ndarray | DaskArray)
 
-@njit
-def _scale_sparse_numba(indptr, indices, data, *, std, mask_obs, clip):
-    for i in numba.prange(len(indptr) - 1):
-        if mask_obs[i]:
-            for j in range(indptr[i], indptr[i + 1]):
-                if clip:
-                    data[j] = min(clip, data[j] / std[indices[j]])
-                else:
-                    data[j] /= std[indices[j]]
+
+@singledispatch
+def clip(x: ArrayLike | _A, *, max_value: float, zero_center: bool = True) -> _A:
```

Comment:
Why do we need `ArrayLike`?

Reply:
For `zappy`?

Reply:
Yeah, this makes `scale` work for things that behave sufficiently like a numpy array.